### DISCUSSION TITLE
Training a universal vocoder

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#252] [Archived]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 10:21am]<br/> <br/> Hello,<br/> <br/> Since training a vocoder takes time and compute, I'd like to train and<br/> contribute a universal vocoder that works for most use cases. *slash* <br/> I have compute but I'm no expert on TTS and I'd like help choosing hyper<br/> parameters and tuning the config file.<br/> <br/> did it with [WaveRNN](https://github.com/erogol/WaveRNN) and it worked<br/> very well. *slash* <br/> I'd like to do the same with faster inference speed to cover more use<br/> cases by using either MelGAN or PWGAN on the same LibriTTS dataset.<br/> <br/> According to my understanding, the sample-rate of the dataset used to<br/> train Tacotron doesn't really matter because it shouldn't affect the mel<br/> spectogram (I'm not so sure about that), the only parameters that should<br/> affect it are :<br/> <br/> should be fixable without retraining)<br/> <br/> And so, still according to my understanding, these are the parameters<br/> that must be shared with all models that use the same vocoder.<br/> <br/> The vocoder's output sample rate shouldn't matter too much but I think<br/> that 16kHz instead of the LibriTTS's 24kHz should give a 33% boost in<br/> inference performance. (I'm not so sure about that as well since PWGAN<br/> and MelGAN are far more parallelised than WaveRNN)<br/> <br/> What do you think about that ? Is it a good idea ? Am I off in my<br/> understanding of the TTS process ?<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/training-a-universal-vocoder]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [August 9, 2020, 10:30am]<br/> <br/> Good idea, thought it might be better to downsample to 22050, since most<br/> TTS training I have seen happens with that sample rate. I also need to<br/> point out that I tried to train a universal vocoder using LibriTTS<br/> **many** times, but unfortunately it never worked well.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 10:36am]<br/> <br/> Could you please tell me what models did you try please and the rough<br/> configs ?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [August 9, 2020, 10:44am]<br/> <br/> I only tried ParallelWaveGAN. The sample rate was 22050. The win length<br/> was 1100 and hop_length was 275. 80 mels and 0 mel_fmin. What I got was<br/> a lot of static and muffled voices. In a paper, it is mentioned it is<br/> much better to train on a much smaller amount of speakers, but with same<br/> amount of speech for each of them. The paper mentioned 6 speakers (3<br/> male and 3 female) with 10 hours for each.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 11:41am]<br/> <br/> Hum ok thanks for the insights. *slash* <br/> So I may go with m-ailabs on French/German/English one speaker f/m each<br/> then... *slash* <br/> Did your config work fine on one speaker ? Do you think MelGAN would<br/> behave better with a high number of speakers ? (I don't think MelGAN's<br/> MOS is sufficient for the tradeoff speed/quality)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 9, 2020, 12:11pm]<br/> <br/> I am able to train PWGAN with LibriTTS better than static noise (still<br/> training). However my initial work indicated that we need to use a<br/> larger model since the original PWGAN model is very tiny and the variety<br/> in the speech requires a stronger model.<br/> <br/> I also try to make the model more sampling rate agnostic by providing a<br/> different up-sampling network for each target sampling rate. And I<br/> provide different sampling rate input for each batch at training.<br/> <br/> you can check the code here<br/> <https://github.com/erogol/TTS_experiments/tree/generic_vocoder><br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 1:10pm]<br/> <br/> Oh ok I get why the sample rate of the TTS training data matters, it's<br/> because for win_length, the unit of the 1024 is sample, so 1024 samples<br/> in 16k is not temporally the same as 24k. So I was wrong, input sr<br/> matters.<br/> <br/> What I don't get is why does the output sample rate of the vocoder<br/> matters, instead of focusing on variable output sample rate, shouldn't<br/> we focus instead on variable input sample rate ? (since different sample<br/> rate produce different Mel Spectograms). *slash* <br/> We could for exemple have the first few layers sr specific by switching<br/> it during trainning. *slash* <br/> Or we could train a small model to do 16k Mel spectogram to 24k Mel<br/> spectogram (somthing like U-net)<br/> <br/> ?<br/> <br/> thanks<br/> <br/> Edit: not u-net because the size of the spectrogram varies with<br/> duration, or windowed<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 1:33pm]<br/> <br/> Ok, just read your code. *slash* <br/> I thought the upsampling was at the end of the network since you said<br/> target sampling rate. Forget what I said about U-net. *slash* <br/> Wouldn't it be simpler for the model to learn if the output sample rate<br/> was fixed ? *slash* <br/> Because if I understand your code correctly, each time it produce a<br/> waveform with the target sr<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 9, 2020, 3:25pm]<br/> <br/> It is another approach to produce a constant sr from a given input.<br/> Maybe you can try this.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 9, 2020, 3:40pm]<br/> <br/> Alright, RN I'm trying converting 16k Mel Spectrograms into 22k ones<br/> with a model to see if without retraining, my Tacotron 16kHz can use<br/> Waveglow 22050Hz as a vocoder. *slash* <br/> If it doesn't work I'm gonna try constant constant sr output.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 9, 2020, 4:41pm]<br/> <br/> I think it's a great start and actually a better approach probably to<br/> solve the sr mismatch unless the user does not like the flexibility of<br/> having multiple sr options.<br/> <br/> Looking forward to see your results<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 9, 2020, 4:47pm]<br/> <br/> I think it also makes sense to train this mel-spec model for multiple<br/> sampling rates like 16k 22.50 and 24. I'd say it'd be hard to convert<br/> from many-to-one without explicitly providing input sr.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 10, 2020, 8:56am]<br/> <br/> Hi quick update,<br/> <br/> Just to test something I tried using the vocoder in the<br/> MultiBand_MelGAN_Example notebook trained in 22050Hz with my Tacotron2<br/> fr 50k 16kHz.<br/> <br/> Here's a sample with griffin_lim. *slash* <br/> .<https://soundcloud.com/julian-weber-8/gl> *slash* <br/> Here's one where I passed the mel_spec without processing *slash* <br/> .<https://soundcloud.com/julian-weber-8/without-pitch-correction> *slash* <br/> and then corrected for pitch with librosa.effects.pitch_shift and noise<br/> reduction *slash* <br/> .<https://soundcloud.com/julian-weber-8/with-pitch-correction> *slash* <br/> And finally, the better sounding one, where I stretched the mel-spec to<br/> the right length with Lanczos resampling. *slash* <br/> .<https://soundcloud.com/julian-weber-8/16ktts-22kmelgan><br/> <br/> This is gonna be my baseline for the model that I'll build although I'm<br/> not too sure that I can do better than that because your vocoder was<br/> trained on a different voice/language<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 13, 2020, 12:39pm]<br/> <br/> Universal PWGAN looks promising after 500k iterations. But I guess it<br/> requires a larger model. I'll this run continue and start another one<br/> with a larger model. Using different upsamle_net for different sampling<br/> rates looks working as well. This model can produce speech in 16khz,<br/> 22050hz and 24khz.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [August 13, 2020, 12:51pm]<br/> <br/> Sounds great. What is the quality you get on the smaller model? Mine<br/> could generalize, but there was a 'zzzzz' sound, and when I used more<br/> speakers, it sounded muffled.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 13, 2020, 1:08pm]<br/> <br/> just a small background noise<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [August 13, 2020, 1:12pm]<br/> <br/> Are you training using the vocoder module on the dev branch? Could you<br/> share your config? If I have time maybe I can try training too<br/> Although I think it is useless seeing as you have reached 500K. Are you<br/> planning to release the model?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> julian.weber<br/> <DATE> [August 14, 2020, 9:23am]<br/> <br/> That's great ! *slash* <br/> Maybe for the background noise, we could try tweaking a bit the loss for<br/> minimum activation in the last layer of the generator ? (That might be a<br/> dumb idea but it's cheap to try since you don't have to retrain the<br/> whole model) *slash* <br/> I'm sorry but I had to work on something else this week and maybe the<br/> beginning of next week. I'll keep you posted if I make any improvements.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 15, 2020, 6:54pm]<br/> <br/> > try tweaking a bit the loss for minimum activation in the last layer<br/> > of the generator<br/> <br/> what do you mean exactly with it?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [August 15, 2020, 6:55pm]<br/> <br/> I stopped the training since the results do not improve. The model is<br/> able to produce good speech for speakers in LibriTTS but some background<br/> noise again for new speakers. The next, I'll try a larger model.<br/> <br/> [**[next page<br/> →](https://discourse.mozilla.org/t/training-a-universal-vocoder/65388?page=2)**<br/> <br/> [Archived Post]<br/>
